src/daft-io/src/object_io.rs

```diff
@@ -108,47 +111,173 @@ pub(crate) trait ObjectSource: Sync + Send {
     async fn ls(
         &self,
         path: &str,
-        delimiter: Option<&str>,
+        delimiter: &str,
+        posix: bool,
         continuation_token: Option<&str>,
+        page_size: Option<i32>,
     ) -> super::Result<LSResult>;
 
-    async fn iter_dir(
+    async fn iter_dir_pages(
         &self,
         uri: &str,
-        delimiter: Option<&str>,
-        _limit: Option<usize>,
-    ) -> super::Result<BoxStream<super::Result<FileMetadata>>> {
+        delimiter: &str,
+        posix: bool,
+        page_size: Option<i32>,
+    ) -> super::Result<BoxStream<super::Result<LSResult>>> {
         let uri = uri.to_string();
-        let delimiter = delimiter.map(String::from);
+        let delimiter = delimiter.to_string();
         let s = stream! {
-            let lsr = self.ls(&uri, delimiter.as_deref(), None).await?;
+            let lsr = self.ls(&uri, delimiter.as_str(), posix, None, page_size).await?;
             let mut continuation_token = lsr.continuation_token.clone();
-            for file in lsr.files {
-                yield Ok(file);
-            }
+            yield Ok(lsr);
 
             while continuation_token.is_some() {
-                let lsr = self.ls(&uri, delimiter.as_deref(), continuation_token.as_deref()).await?;
+                let lsr = self.ls(&uri, delimiter.as_str(), posix, continuation_token.as_deref(), page_size).await?;
                 continuation_token = lsr.continuation_token.clone();
-                for file in lsr.files {
-                    yield Ok(file);
+                yield Ok(lsr);
+            }
+        };
+        Ok(s.boxed())
+    }
+
+    async fn iter_dir(
```

Comment:
NOTE: This is changed to return `LSResult` in #1460 

For the sake of easier review of this PR

Reply:
(Removed, no need to paginate)